src/masternode-budget.cpp

```diff
@@ -457,18 +457,8 @@ bool CBudgetManager::AddProposal(CBudgetProposal& budgetProposal)
 
 void CBudgetManager::CheckAndRemove()
 {
-    int nHeight = 0;
-
-    // Add some verbosity once loading blocks from files has finished
-    {
-        TRY_LOCK(cs_main, locked);
-        if ((locked) && (chainActive.Tip() != NULL)) {
-            CBlockIndex* pindexPrev = chainActive.Tip();
-            if (pindexPrev) {
-                nHeight = pindexPrev->nHeight;
-            }
-        }
-    }
+    int nHeight = WITH_LOCK(cs_main, return chainActive.Height());
+    if (nHeight <= 0) return;
```

Comment:
let's better remove these two lines.
The check is not needed for the internal procedure, should be checked on the caller's side if needed and most importantly, can be a deadlocking cause.

The callers are:

 `CBudgetManager::NewBlock` --> which doesn't need it. Method only called when a new block arrives. Height will be never 0.

`CBudgetDB::Read` --> only called during init. A check flag can be set before the `LOCK(objToLoad.cs);` line to call or not `CheckAndRemove` below.

`checkbudgets` --> RPC command. Can be checked before calling the method.

Reply:
Good Idea.
For the DB read, we can just modify the `fDryRun` flag in init. For the RPC function, I actually think it's better to check before for complete mnsync anyway.
Updated this in e09f3164d78b52091f9158e6d14322cb48c79544